LazyAlchemyLearner.lua

```diff
@@ -142,41 +187,164 @@ local function getItemLinkFromItemId(itemId)
-			GetItemLinkReagentTraitInfo(getItemLinkFromItemId(comboInfo[i][2]), j)
-		end
-		if remainingSolvent == 0 then
-			local solventProficiency = GetNonCombatBonus(NON_COMBAT_BONUS_ALCHEMY_LEVEL)
-			if solventProficiency == 0 then
-				solventProficiency = 1
+		
+			--[[
+				k = known status of the reagant's trait {true/false}
+				n = name of the trait, currently the name wil be nil if the trait is unknown
+			--]]
+			local k1, n1 = GetItemLinkReagentTraitInfo(getItemLinkFromItemId(combos[i][1]), j)
+			local k2, n2 = GetItemLinkReagentTraitInfo(getItemLinkFromItemId(combos[i][2]), j)
+
+			-- as a precaution, we'll set the name of the trait to nil if it's not known, since this is the behaviour we expect in the next step
+			if not k1 then
+				n1 = nil
+			end
+
+			if not k2 then
+				n2 = nil
+			end
+			
+			--we count the amount of known matching traits of each reagant
+			if Contains(theoreticalMatchingTraits, n1) then
+				amountMatchingReagant1 = amountMatchingReagant1 + 1
 			end
-			remainingSolvent, solvent, position = getSolvent(solventProficiency, position-1)
 			
-		else
+			if Contains(theoreticalMatchingTraits, n2) then
+				amountMatchingReagant2 = amountMatchingReagant2 + 1
+			end
+		end
+		
+		-- if the amount of actual known, matching traits is not the same as the theoretical amount of matching traits known will be false
+		known = amountMatchingReagant1 == #theoreticalMatchingTraits and amountMatchingReagant2 == #theoreticalMatchingTraits
 		
-			if not known then -- 75358
-				remainingSolvent = remainingSolvent - 1
+		--Leaving this here for debug purposes for now
+		 local reagantName1 = GetItemLinkName(getItemLinkFromItemId(combos[i][1]))
+		 local reagantName2 = GetItemLinkName(getItemLinkFromItemId(combos[i][2]))
+		
+		--if unknown traits detected we will attempt to queue a potion
+		if not known then
+		
+			-- we check if there are any solvents left of the currently selected solvent, if not we automatically get the next best solvent
+			if (remainingSolvent == 0) then
+				--decide the solvent we are going to use
+				remainingSolvent, solvent, position = getSolvent(GetNonCombatBonus(NON_COMBAT_BONUS_ALCHEMY_LEVEL), position)
+	
+				if(remainingSolvent == nil) then
+					d("You need solvents to craft potions")
+					return queued
+				end
+			end
+		
+			-- we reduce the amount of available solvents
+			remainingSolvent = remainingSolvent - 1
+			
+			-- Check if we have enough reagent available, if not we output to chat and skip the potion
+			local canCraftPotion = true
+			
+			-- check availability of reagant 1
+			if(reagentAmounts[combos[i][1]] == 0) then
+				reagentAmounts[combos[i][1]]  = GetAvailableReagent(combos[i][1])
+			end
+
+			if(reagentAmounts[combos[i][1]]  == nil) then
+				--skip this potion
+				d("Lazy Alchemy Learner: Not enough " .. tostring(reagantName1) .. " to learn all its traits.")
+				canCraftPotion = false
+			end
```

Comment:
```suggestion
			-- check availability of reagant 1
			if not reagentAmounts[combos[i][1]] then
				reagentAmounts[combos[i][1]]  = GetAvailableReagent(combos[i][1])
			end

			if reagentAmounts[combos[i][1]]  == 0 then
				--skip this potion
				d("Lazy Alchemy Learner: Not enough " .. tostring(reagantName1) .. " to learn all its traits.")
				canCraftPotion = false
			end
```

Bit of a multi-part suggestion. With this, should also change GetAvailableReagent to return 0 as a default, and initialize reagentAmounts as an empty table, instead of with initial values of 0.

Reply:
I think this is actually mandatory, was thinking about it last night. With the current inplementation, once availableReagant becomes 0 it will just recalculate. Will look at it this evening before merging.

Reply:
Good point, it won't get into a loop at least, but it is unnecessary calculation. (Not that we're doing much but... still)